README.rst
==============================================================================
A robot framework library for Django.
==============================================================================

.. image:: https://travis-ci.org/kitconcept/robotframework-djangolibrary.svg?branch=master
    :target: https://travis-ci.org/kitconcept/robotframework-djangolibrary

.. image:: https://img.shields.io/pypi/status/robotframework-djangolibrary.svg
    :target: https://pypi.python.org/pypi/robotframework-djangolibrary/
    :alt: Egg Status

.. image:: https://img.shields.io/pypi/v/robotframework-djangolibrary.svg
    :target: https://pypi.python.org/pypi/robotframework-djangolibrary/
    :alt: Latest Version

.. image:: https://img.shields.io/pypi/l/robotframework-djangolibrary.svg
    :target: https://pypi.python.org/pypi/robotframework-djangolibrary/
    :alt: License

|

.. image:: https://raw.githubusercontent.com/kitconcept/robotframework-djangolibrary/master/kitconcept.png
   :alt: kitconcept
   :target: https://kitconcept.com/


Introduction
------------

DjangoLibrary is a web testing library to test Django with Robot Framework.
It uses SeleniumLibrary to run tests against a real browser instance.

The library will automatically start and stop your Django instance while running the tests.
It also comes with serveral autologin keywords that allow you to login different users during your tests, without the need to actually access the login page.

DjangoLibrary is currently tested against Django 1.8.x, 1.9.x, 1.11.x with SQLite and Postgres on Python 2.7 and 3.6.


Documentation
-------------

`Robot Framework Django Library Keyword Documentation`_


Installation
------------

Install robotframework-djangolibrary with pip::

  $ pip install robotframework-djangolibrary

In order to be able to use DjangoLibrary's `Autologin`, `FactoryBoy`, or
`QuerySet` keywords you have to add the corresponding middleware classes to
your MIDDLEWARE_CLASSES in yoursettings.py::

  MIDDLEWARE_CLASSES = (
      ...
      'django.contrib.auth.middleware.AuthenticationMiddleware',
      'DjangoLibrary.middleware.AutologinAuthenticationMiddleware',
      'DjangoLibrary.middleware.FactoryBoyMiddleware',
      'DjangoLibrary.middleware.QuerySetMiddleware',
  )

.. DANGER::
   Make sure that you add those middlewares only to your test setup and
   NEVER to your deployment! The Autologin middleware just checks for a
   'autologin' cookie and then authenticates and login ANY user.


First Robot Test
----------------

In order to write your first robot test, make sure that you include SeleniumLibrary and DjangoLibrary. Create a test.robot file with the
following content::

  *** Variables ***

  ${HOSTNAME}             127.0.0.1
  ${PORT}                 55001
  ${SERVER}               http://${HOSTNAME}:${PORT}/
  ${BROWSER}              firefox


  *** Settings ***

  Documentation   Django Robot Tests
  Library         SeleniumLibrary  timeout=10  implicit_wait=0
  Library         DjangoLibrary  ${HOSTNAME}  ${PORT}  path=mysite/mysite  manage=mysite/manage.py  settings=mysite.settings
  Suite Setup     Start Django and open Browser
  Suite Teardown  Stop Django and close Browser


  *** Keywords ***

  Start Django and open Browser
    Start Django
    Open Browser  ${SERVER}  ${BROWSER}

  Stop Django and close browser
    Close Browser
    Stop Django


  *** Test Cases ***

  Scenario: As a visitor I can visit the django default page
    Go To  ${SERVER}
    Wait until page contains element  id=explanation
    Page Should Contain  It worked!
    Page Should Contain  Congratulations on your first Django-powered page.


License
-------

Copyright kitconcept GmbH.

Distributed under the terms of the Apache License 2.0, robotframework-djangolibrary is free and Open Source software.


Contribute
----------

- `Source code at Github <https://github.com/kitconcept/robotframework-djangolibrary>`_
- `Issue tracker at Github <https://github.com/kitconcept/robotframework-djangolibrary/issues>`_


Support
-------

If you are having issues, `please let us know <https://github.com/kitconcept/robotframework-djangolibrary/issues>`_. If you require professional support feel free to contact us at `info@kitconcept.com. <mailto:info@kitconcept.com>`_


Run Tests
---------

Then you can run the test with robot::

  $ robot test.robot

The output should look like this::

  ==============================================================================
  Test :: Django Robot Tests
  ==============================================================================
  Scenario: As a visitor I can visit the django default page            | PASS |
  ------------------------------------------------------------------------------
  Test :: Django Robot Tests                                            | PASS |
  1 critical test, 1 passed, 0 failed
  1 test total, 1 passed, 0 failed
  ==============================================================================
  Output:  /home/timo/workspace/prounix/robotframework-djangolibrary/output.xml
  Log:     /home/timo/workspace/prounix/robotframework-djangolibrary/log.html
  Report:  /home/timo/workspace/prounix/robotframework-djangolibrary/report.html


Test Isolation
--------------

robotframework-djangolibrary does not provide isolation between tests by
default. This means if you add an object to the database in a test, this
object will be present in the next test as well. You need to cleanup
yourself in order to have a proper isolation between the tests. You can use
the robotframework "Test Teardown" call to call the "Clear DB" keyword after
each test::

  *** Settings ***

  Library         SeleniumLibrary  timeout=10  implicit_wait=0
  Library         DjangoLibrary  ${HOSTNAME}  ${PORT}  path=mysite/mysite  manage=mysite/manage.py  settings=mysite.settings  db=mysite/db.sqlite3
  Suite Setup     Start Django and open Browser
  Suite Teardown  Stop Django and close Browser
  Test Teardown   Clear DB


Development
-----------

Checkout repository from github::

  $ git clone https://github.com/kitconcept/robotframework-djangolibrary.git

Create a virtual Python environment::

  $ cd robotframework-djangolibrary/
  $ virtualenv .py27
  $ source .py27/bin/activate

Install robotframework-djangolibrary in development mode::

  $ python setup.py develop

Install the requirements::

  $ pip install -r requirements.txt

Run Unit/Integration-Tests::

  $ py.test mysite/

Run Acceptance Tests::

  $ robot DjangoLibrary/tests/

.. _`Robot Framework Django Library Keyword Documentation`: https://kitconcept.github.io/robotframework-djangolibrary/


CHANGES.rst

3.1.1 (unreleased)
------------------

Bug Fixes:

- Support models that have a primary key called something other than "id".


3.1.0 (2018-12-24)
------------------

New Features:

- Python 3.7 support.
  [timo]


3.0.0 (2018-11-26)
------------------

Breaking Changes:

- Depend on SeleniumLibrary instead of Selenium2Library.
  The name changed from Selenium2Library to SeleniumLibrary.
  See https://github.com/robotframework/SeleniumLibrary/blob/master/docs/SeleniumLibrary-3.0.0.rst#name-changed-from-selenium2library-to-seleniumlibrary
  for details.
  [timo]

New Features:

- Python 3.6 support (earlier versions most likely do support Python 3.6 as well, we just did not test it so far).
  [timo]


2.0.2 (2018-03-03)
------------------

Bugfixes / Minor Changes:

- Fix Pypi markup.
  [timo]


2.0.1 (2018-03-03)
------------------

Bugfixes / Minor Changes:

- Fix outdated documentation.
  [timo]


2.0 (2017-09-20)
----------------

Breaking Changes:

- Remove deprecated method from internal startup flow.
  [cdvv7788]

New Features:

- Use AUTH_MODEL to create user instead of django's default
  [cdvv7788]

- Upgrade FactoryBoy to version 2.8.1
  [amarandon]

- Add Django 1.10.7 and 1.11.1 support.
  [timo]

- LICENSE.txt added.
  [timo]

Bugfixes / Minor Changes:

- Move tests to use Chrome.
  [timo]

- Upgrade tests to Django 1.11.5, 1.10.8, 1.9.13, 1.8.18.
  [timo]

- Add Django 1.11 and 1.10 to setup.py.
  [timo]


1.2 (2016-07-08)
----------------

New Features:

- Make Factory Boy keyword return 'pk' attribute.
  [timo]


1.1 (2016-07-07)
----------------

New Features:

- Add QuerySet keyword.
  [timo]

- Make it possible to override subfactories when using the `FactoryBoy`
  keyword.
  [timo]

Bugfixes:

- Use Django's model_to_dict method to serialize the response objects for the
  factory_boy keyword.
  [timo]


1.0 (2016-06-30)
----------------

- Re-release 1.0a6 as 1.0.
  [timo]


1.0a6 (2016-04-29)
------------------

New Features:

- Python 3 compatibility. Note that the latest offical release of
  robotframework-selenium2library is currently not compatible with Python 3.
  See https://github.com/HelioGuilherme66/robotframework-selenium2library/releases for a working pre-release and details.
  [timo]

- Support for Postgres added. All Django database backends should work.
  We test SQLite and Postgres only though.
  [timo]

- Add 'Factory Boy' keyword. This allows us to use factory_boy factories in
  Robot Framework tests.
  [timo]

Breaking Changes:

- Drop Django 1.7.x support. We test and support Django 1.8.x and 1.9.x.
  [timo]

- Change 'Clear DB' implementation to use "python manage.py flush" instead of
  deleting and re-building the database.
  [timo]

- Remove 'Debug' and 'Pause' keywords. The 'Debug' keyword, which is
  provided by robotframework-debuglibrary is sufficient.
  [timo]


1.0a5 (2016-02-11)
------------------

- Make middleware part Python 3 compatible.

  robotframework-djangolibrary is still not compatible with Python 3 because
  robotframework-selenium2library does not work with Python 3 yet. Though, you
  can install robotframwork-djangolibrary on Python 3 with "pip install
  robotframework-djangolibrary --no-deps" and then run your tests with
  Python 2.7.
  [timo]

- Add 'Framework :: Robot Framework' classifier to setup.py.
  [timo]


1.0a4 (2016-02-05)
------------------

- Use 'migrate' instead of 'syncdb' for Django > 1.7.x.
  [timo]


1.0a3 (2015-09-28)
------------------

- Add list_classifiers to setup.py.
  [timo]

- Fix user creation and startup. This fixes #3.
  [MatthewWilkes]


1.0a2 (2015-06-25)
------------------

- Remove Django and zest.releaser from requirements.txt. This fixes #2.
  [timo]


1.0a1 (2015-06-24)
------------------

- Initial release.
  [timo]


DjangoLibrary/__init__.py
# -*- coding: utf-8 -*-
from robot.api import logger
from robot.libraries.BuiltIn import BuiltIn
from warnings import warn

import base64
import json
import os
import requests
import signal
import six
import subprocess

__version__ = '1.0'
ROBOT_LIBRARY_DOC_FORMAT = 'reST'


def safe_bytes(str):
    """Returns bytes on Py3 and a string on Py2."""
    if six.PY3:
        return bytes(str, 'utf-8')
    else:
        return str


def safe_utf8(string):
    """Returns bytes on Py3 and an utf-8 encoded string on Py2."""
    if six.PY2:
        return string.encode("utf-8")
    else:
        return string


class DjangoLibrary:
    """DjangoLibrary is a web testing library to test Django with Robot
    Framework.

    It uses SeleniumLibrary to run tests against a real browser instance.

    *Before running tests*

    Prior to running test cases using DjangoLibrary, DjangoLibrary must be
    imported (together with SeleniumLibrary) into your Robot test suite
    (see `importing` section), and the SeleniumLibrary 'Open Browser' keyword
    must be used to open a browser to the desired location.
    """

    django_pid = None

    # TEST CASE => New instance is created for every test case.
    # TEST SUITE => New instance is created for every test suite.
    # GLOBAL => Only one instance is created during the whole test execution.
    ROBOT_LIBRARY_SCOPE = 'TEST SUITE'

    def __init__(self, host="0.0.0.0", port=8000, path='mysite/mysite',
                 manage='mysite/manage.py', settings='mysite.settings',
                 db=None):
        """Django2Library can be imported with optional arguments.

        `host` is the hostname of your Django instance. Default value is
        '127.0.0.1'.

        `port` is the port number of your Django instance. Default value is
        8000.

        `path` is the path to your Django instance.

        `manage` is the path to your Django instance manage.py.

        `settings` is the path to your Django instance settings.py.

        `db` is deprecated. Please don't use it.

        Examples:
        | Library | SeleniumLibrary | timeout=15        | implicit_wait=0.5  | # Sets default timeout to 15 seconds and the default implicit_wait to 0.5 seconds. |  # noqa
        | Library | DjangoLibrary    | 127.0.0.1         | 55001              | path=mysite/mysite | manage=mysite/manage.py | settings=mysite.settings | db=mysite/db.sqlite3 | # Sets default hostname to 127.0.0.1 and the default port to 55001.                |  # noqa
        """
        self.host = host
        self.port = port
        self.path = os.path.realpath(path)
        self.manage = os.path.realpath(manage)
        self.settings = settings
        if db:
            warn(
                "Using the DjangoLibrary 'db' parameter is deprecated. " +
                "Use the 'settings' parameter instead to set a " +
                "database connection."
            )

    def manage_makemigrations(self):
        """Create migrations by running 'python manage.py makemigrations'."""
        args = [
            'python',
            self.manage,
            'makemigrations',
        ]
        subprocess.call(args)

    def manage_migrate(self):
        """Execute migration by running 'python manage.py migrate'."""
        args = [
            'python',
            self.manage,
            'migrate',
            '--settings=%s' % self.settings,
        ]
        subprocess.call(args)

    def manage_flush(self):
        """Clear database by running 'python manage.py flush'."""
        args = [
            'python',
            self.manage,
            'flush',
            '--noinput',
            '--settings=%s' % self.settings,
        ]
        subprocess.call(args)

    def clear_db(self):
        """Clear database. This is a legacy keyword now. Use 'Manage Flush'
           instead.
        """
        warn(
            "The DjangoLibrary 'clear_db' keyword is deprecated. " +
            "Use the 'manage_flush' keyword instead."
        )
        self.manage_flush()

    def create_user(self, username, email, password, **kwargs):
        """Create a regular Django user in the default auth model.

        The `Create User` keyword allows to provide additional arguments that
        are passed directly to the Djange create_user method (e.g.
        "is_staff=True")."""

        to_run = """
from django.contrib.auth import get_user_model
user = get_user_model().objects.create_user(
    '{0}',
    email='{1}',
    password='{2}',
)
user.is_superuser = '{3}'
user.is_staff = '{4}'
user.save()""".format(
            safe_utf8(username),
            safe_utf8(email),
            safe_utf8(password),
            kwargs.get('is_superuser', False),
            kwargs.get('is_staff', False),
        )
        args = [
            'python',
            self.manage,
            'shell',
            '--plain',
            '--settings=%s' % self.settings,
        ]

        django = subprocess.Popen(
            args,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE
        )

        django.communicate(safe_bytes(to_run))

    def create_superuser(self, username, email, password):
        """Create a Django superuser in the default auth model."""
        self.create_user(username, email, password,
                         is_superuser=True, is_staff=True)

    def start_django(self):
        """Start the Django server."""
        self.manage_flush()
        self.manage_makemigrations()
        self.manage_migrate()
        logger.console("-" * 78)
        args = [
            'python',
            self.manage,
            'runserver',
            '%s:%s' % (self.host, self.port),
            '--nothreading',
            '--noreload',
            '--settings=%s' % self.settings,
        ]

        self.django_pid = subprocess.Popen(
            args,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE
        ).pid
        logger.console(
            "Django started (PID: %s)" % self.django_pid,
        )
        logger.console("-" * 78)

    def stop_django(self):
        """Stop the Django server."""
        os.kill(self.django_pid, signal.SIGKILL)
        logger.console(
            "Django stopped (PID: %s)" % self.django_pid,
        )
        logger.console("-" * 78)

    def autologin_as(self, username, password):
        """Autologin as Django user.

        DjangoLibrary comes with a Django middleware component that allows the
        autologin_as keyword to set an 'autologin' cookie that the
        middleware uses to authenticate and login the user in Django.

        If you want to use the autlogin_as keyword you have to add
        'DjangoLibrary.middleware.AutologinAuthenticationMiddleware' to the
        MIDDLEWARE_CLASSES right after the default AuthenticationMiddleware
        in your settings.py::

            MIDDLEWARE_CLASSES = (
                ...
                'django.contrib.auth.middleware.AuthenticationMiddleware',
                'DjangoLibrary.middleware.AutologinAuthenticationMiddleware',
            )

        *Warning*

        Make sure that you add this middleware only to your test setup and
        NEVER to your deployment!

        See https://github.com/kitconcept/robotframework-djangolibrary/blob/master/DjangoLibrary/tests/test_autologin.robot  # noqa
        for examples how to use the `Autologin As` keyword.

        """
        if six.PY2:
            username = username.encode('utf-8')
            password = password.encode('utf-8')
        # encode autologin cookie value as base64
        autologin_cookie_value = base64.b64encode(
            safe_bytes("%s:%s" % (username, password))
        )

        selenium2lib = BuiltIn().get_library_instance('SeleniumLibrary')
        # XXX: The 'Add Cookie' keywords does not work with Firefox, therefore
        # we have to add the cookie with js here. A bug has been filed:
        # https://github.com/rtomac/robotframework-selenium2library/issues/273
        # selenium2lib.add_cookie(
        #     "autologin",
        #     "%s:%s" % (username, password),
        #     path="/",
        #     domain="localhost",
        # )

        if six.PY3:
            selenium2lib.execute_javascript(
                "document.cookie = 'autologin=%s;path=/;domain=localhost;';" %
                autologin_cookie_value.decode('utf-8')
            )
        else:
            selenium2lib.execute_javascript(
                "document.cookie = 'autologin=%s;path=/;domain=localhost;';" %
                autologin_cookie_value
            )

        # autologin_cookie = selenium2lib.get_cookie_value('autologin')
        # assert autologin_cookie == "%s:%s" % (username, password)
        # cookies = selenium2lib.get_cookies()
        # assert cookies == u"autologin=%s:%s" % (username, password)

    def autologin_logout(self):
        """Logout a user that has been logged in by the autologin_as keyword.
        """
        selenium2lib = BuiltIn().get_library_instance('SeleniumLibrary')
        selenium2lib.execute_javascript(
            "document.cookie = 'autologin=;path=/;domain=localhost;';"
        )

    def factory_boy(self, factory, **kwargs):
        """Create content objects in the Django database with Factory Boy.
        See https://factoryboy.readthedocs.org for more details.

        Arguments:

        `factory` is a required argument and should contain the full path to
        your factory boy factory class (e.g.
        "FactoryBoy mysite.polls.factories.PollFactory").

        The `Factory Boy` keyword allows to provide additional arguments that
        are passed directly to the Factory Boy Factory class
        (e.g. "FactoryBoy mysite.polls.factories.PollFactory pollname='mypoll'").

        You can also override subfactories by using the double-underscore
        field lookup (https://docs.djangoproject.com/en/1.9/topics/db/queries/#field-lookups)  #noqa
        together with the `pk` lookup shortcut (https://docs.djangoproject.com/en/1.9/topics/db/queries/#the-pk-lookup-shortcut)
        (e.g. "Factory Boy  bookstore.factories.BookFactory ... author__pk=1")

        See https://github.com/kitconcept/robotframework-djangolibrary/blob/master/DjangoLibrary/tests/test_factory_boy.robot  # noqa
        for examples how to use the `Factory Boy` keyword.

        """
        url = 'http://{}:{}'.format(
            self.host,
            self.port
        )
        payload = {
            'FACTORY_BOY_MODEL_PATH': factory,
            'FACTORY_BOY_ARGS': json.dumps(kwargs)
        }
        response = requests.get(url, params=payload)
        if response.status_code == 201:
            return response.json()
        status_code_400 = response.status_code == 400
        type_json = response.headers.get('Content-Type') == 'application/json'
        if status_code_400 and type_json:
            msg = response.json().get('error', '')
            traceback = response.json().get('traceback', '')
            if traceback:
                msg = msg + '\n\n' + traceback
            raise requests.exceptions.HTTPError(msg, response=response)
        return response.raise_for_status()

    def query_set(self, model, **kwargs):
        """Query the Django ORM.

        Returns a QuerySet object. See https://docs.djangoproject.com/en/1.9/topics/db/queries/#retrieving-objects for details.  # noqa

        Arguments:

        `model` is a required argument and should contain the full path to
        your Django model class (e.g. "django.contrib.auth.models.User").

        The `QuerySet` keyword allows to provide additional arguments that
        are passed as filter arguments
        (e.g. "django.contrib.auth.models.User  username=john").
        If no additonal argument is provided `QuerySet will just return all
        objects that exists for that model.

        `limit` limits the number of results,
        e.g. "django.contrib.auth.models.User  limit=10" will return 10 results
        max.
        Limit is an optional argument that maps 1:1 to the QuerySet limit
        argument. See https://docs.djangoproject.com/en/1.9/topics/db/queries/#limiting-querysets  # noqa
        for details.

        `offset` can be used in combination with `limit` to set an offset,
        e.g. "django.contrib.auth.models.User  offeset=5  limit=10" will return
        5 results while omitting the first 5 results.
        See https://docs.djangoproject.com/en/1.9/topics/db/queries/#limiting-querysets  # noqa
        for details.

        See https://github.com/kitconcept/robotframework-djangolibrary/blob/master/DjangoLibrary/tests/test_query_set.robot  # noqa
        for examples how to use the `Query Set` keyword.

        """
        url = 'http://{}:{}'.format(
            self.host,
            self.port
        )
        payload = {
            'MODEL_PATH': model,
            'QUERY_ARGS': json.dumps(kwargs)
        }
        response = requests.get(url, params=payload)
        if response.status_code == 200:
            return response.json()
        status_code_400 = response.status_code == 400
        type_json = response.headers.get('Content-Type') == 'application/json'
        if status_code_400 and type_json:
            msg = response.json().get('error', '')
            traceback = response.json().get('traceback', '')
            if traceback:
                msg = msg + '\n\n' + traceback
            raise requests.exceptions.HTTPError(msg, response=response)
        return response.raise_for_status()
